File: sdk/python/packages/flet-flashlight/src/flutter/flet_flashlight/lib/src/flashlight.dart
import 'package:flet/flet.dart';
import 'package:flutter/widgets.dart';
import 'package:torch_light/torch_light.dart';

class FlashlightControl extends FletService {
  FlashlightControl({required super.control});

  @override
  void init() {
    super.init();
    debugPrint("Flashlight(${control.id}).init: ${control.properties}");
    control.addInvokeMethodListener(_invokeMethod);
  }

  Future<dynamic> _invokeMethod(String name, dynamic args) async {
    debugPrint("Flashlight.$name($args)");
    if (isMobilePlatform()) {
      switch (name) {
        case "on":
          await TorchLight.enableTorch();
        case "off":
          await TorchLight.disableTorch();
        case "is_available":
          return await TorchLight.isTorchAvailable();
        default:
          throw Exception("Unknown Flashlight method: $name");
      }
    } else {
      throw Exception(
          "Flashlight control is supported only on Android and iOS devices.");
    }
  }

  @override
  void dispose() {
    debugPrint("Flashlight(${control.id}).dispose");
    control.removeInvokeMethodListener(_invokeMethod);
    super.dispose();
  }
}

File: sdk/python/packages/flet-flashlight/src/flet_flashlight/flashlight.py
import flet as ft

__all__ = ["Flashlight"]


@ft.control("Flashlight")
class Flashlight(ft.Service):
    """
    A control to use FlashLight. Works on iOS and Android.
    """

    async def turn_on(self):
        """
        Turns the flashlight on.
        """
        await self._invoke_method("on")

    async def turn_off(self):
        """
        Turns the flashlight off.
        """
        await self._invoke_method("off")

    async def is_available(self):
        """
        Checks if the flashlight is available on the device.
        """
        return await self._invoke_method("is_available")
